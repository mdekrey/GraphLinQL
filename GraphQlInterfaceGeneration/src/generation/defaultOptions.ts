import { Options } from "./Options";

export const defaultOptions: Options = {
  namespace: "GraphLinqQL.Interfaces",
  useNullabilityIndicator: true,
  scalarTypes: {
    Int: { csharpNullable: false, csharpType: "int" },
    Float: { csharpNullable: false, csharpType: "double" },
    String: { csharpNullable: true, csharpType: "string" },
    Boolean: { csharpNullable: false, csharpType: "bool" }
  },
  using: [
    "GraphLinqQL",
    "GraphLinqQL.Introspection",
    "System",
    "System.Collections",
    "System.Collections.Generic",
    "System.Collections.Immutable",
    "System.Linq"
  ],
  deserializer: "System.Text.Json.JsonSerializer.Deserialize",
  introspection: false
};
