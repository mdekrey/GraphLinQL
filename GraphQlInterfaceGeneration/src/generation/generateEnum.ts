import { GraphQLEnumType, GraphQLEnumValue } from "graphql";
import { Options } from "./Options";
import { getTypeName } from "./getTypeName";
import { getEnumValueName } from "./getEnumValueName";

export function generateEnum(object: GraphQLEnumType, options: Options) {
  const typeName = getTypeName(object.name, options);

  return `${
    object.description
      ? `
/// <summary>
/// ${object.description.split("\n").join("\n/// ")}
/// </summary>`
      : `
`
  }
public enum ${typeName}
{
    ${object.getValues().map(value => generateEnumValue(value, options)).join(`,
    `)}
}
`;
}

function generateEnumValue(value: GraphQLEnumValue, options: Options) {
  return (
    (value.description &&
      `
    /// <summary>
    /// ${value.description.split("\n").join("\n    /// ")}
    /// </summary>
    `) + getEnumValueName(value.name, options)
  );
}
