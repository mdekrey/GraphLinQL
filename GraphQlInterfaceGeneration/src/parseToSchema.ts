import { parse, buildASTSchema, GraphQLSchema } from "graphql";

export function parseToSchema(graphQlSchema: string): GraphQLSchema {
  const astDocument = parse(graphQlSchema);

  const backcompatOptions = { commentDescriptions: true };
  const schema = buildASTSchema(astDocument, backcompatOptions);

  return schema;
}
