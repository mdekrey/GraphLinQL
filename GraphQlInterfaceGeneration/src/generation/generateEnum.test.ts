import { parseToSchema } from "../parseToSchema";
import { getStarWarsGraphQlSchemaV3 } from "../resources.test";
import { GraphQLEnumType } from "graphql";
import { defaultOptions } from "./defaultOptions";
import { generateEnum } from "./generateEnum";

it("can generate v3 enum", async function() {
  const schema = await getSampleSchema(getStarWarsGraphQlSchemaV3);
  const type = schema.getType("Episode") as GraphQLEnumType;
  expect(generateEnum(type, defaultOptions)).toMatchSnapshot();
});

async function getSampleSchema(getSchema: () => Promise<string>) {
  const schema = parseToSchema(await getSchema());
  return schema;
}
