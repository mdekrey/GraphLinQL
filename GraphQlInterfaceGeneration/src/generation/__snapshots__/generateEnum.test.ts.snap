// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can generate v3 enum 1`] = `
"
/// <summary>
/// The episodes in the Star Wars trilogy
/// </summary>
public enum Episode
{
    
    /// <summary>
    /// Star Wars Episode IV: A New Hope, released in 1977.
    /// </summary>
    NEWHOPE,
    
    /// <summary>
    /// Star Wars Episode V: The Empire Strikes Back, released in 1980.
    /// </summary>
    EMPIRE,
    
    /// <summary>
    /// Star Wars Episode VI: Return of the Jedi, released in 1983.
    /// </summary>
    JEDI
}
"
`;
